homeassistant/components/binary_sensor/trend.py

```diff
@@ -65,30 +84,40 @@ def setup_platform(hass, config, add_devices, discovery_info=None):
 class SensorTrend(BinarySensorDevice):
     """Representation of a trend Sensor."""
 
-    def __init__(self, hass, device_id, friendly_name,
-                 target_entity, attribute, device_class, invert):
+    def __init__(self, hass, device_id, friendly_name, entity_id,
+                 attribute, device_class, invert, sample_duration,
+                 min_gradient):
         """Initialize the sensor."""
         self._hass = hass
         self.entity_id = generate_entity_id(
             ENTITY_ID_FORMAT, device_id, hass=hass)
         self._name = friendly_name
-        self._target_entity = target_entity
+        self._entity_id = entity_id
         self._attribute = attribute
         self._device_class = device_class
         self._invert = invert
+        self._sample_duration = sample_duration
+        self._min_gradient = min_gradient
+        self._gradient = None
         self._state = None
-        self.from_state = None
-        self.to_state = None
+        self.samples = []
 
         @callback
         def trend_sensor_state_listener(entity, old_state, new_state):
             """Handle the target device state changes."""
-            self.from_state = old_state
-            self.to_state = new_state
-            hass.async_add_job(self.async_update_ha_state(True))
-
-        track_state_change(hass, target_entity,
-                           trend_sensor_state_listener)
+            try:
+                if self._attribute:
+                    state = new_state.attributes.get(self._attribute)
+                else:
+                    state = new_state.state
+                if state != STATE_UNKNOWN:
+                    now = datetime.now(timezone.utc).timestamp()
+                    self.samples.append((now, float(state)))
+                    hass.async_add_job(self.async_update_ha_state(True))
+            except (ValueError, TypeError) as ex:
+                _LOGGER.error(ex)
+
+        track_state_change(hass, entity_id, trend_sensor_state_listener)
```

Comment:
Move that into `async_added_to_hass` coro (need to be create here) and do that `self.hass.helpers.event.async_track_state_change...`